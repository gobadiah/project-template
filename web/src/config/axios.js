import cookie from 'cookie';

export default (args) => {
  const { headers = {}, req, api } = args || {};
  if (req) {
    const cookies = Object.keys(req.cookies).map(key => cookie.serialize(key, req.cookies[key])).join('; ');
    headers.Cookie = cookies;
  }
  return {
    baseUrl: api,
    withCredentials: true,
    headers,
  };
};
