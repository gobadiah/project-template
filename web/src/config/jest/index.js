const path = require('path');

module.exports = {
  rootDir: path.join(__dirname, '../../..'),
  testEnvironment: 'node',
  modulePathIgnorePatterns: [
    '<rootDir>/dist',
  ],
  testPathIgnorePatterns: [
    '/node_modules/',
    '/dist/',
  ],
  setupFiles: [
    '<rootDir>/src/config/jest/setup.js',
  ],
  setupTestFrameworkScriptFile: '<rootDir>/src/config/jest/setup-test-framework.js',
  snapshotSerializers: [
    'enzyme-to-json/serializer',
  ],
  coverageReporters: [
    'json',
    'lcov',
    'text',
    'html',
  ],
  coveragePathIgnorePatterns: [
    '/node_modules/',
    '/dist/',
  ],
  coverageDirectory: '<rootDir>/reports/coverage',
  collectCoverageFrom: [
    '<rootDir>/src/**/*.js',
    '!<rootDir>/src/.next/**',
  ],
};
