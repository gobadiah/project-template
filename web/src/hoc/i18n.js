import i18n from '~/config/i18n';

export default ({ req, namespaces }) => new Promise((resolve) => {
  if (req) {
    resolve(i18n.getInitialProps(req, namespaces));
  } else {
    resolve({});
  }
});
