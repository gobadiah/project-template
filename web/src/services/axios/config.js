import cookie from 'cookie';

import generalConfig from '~/config';

export default ({ headers = {}, req } = {}) => {
  const config = {
    baseUrl: generalConfig.api,
    withCredentials: true,
    headers,
  };
  if (req) {
    const cookies = Object.keys(req.cookies).map(key =>
      cookie.serialize(key, req.cookies[key])).join('; ');
    config.headers.Cookie = cookies;
  }
  return config;
};
