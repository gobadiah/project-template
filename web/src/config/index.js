import axios from './axios';

const port = parseInt(process.env.PORT, 10) || 3000;
const web = process.env.WEB_URL || `http://127.0.0.1:${port}`;
const api = process.browser ? `${web}/api` : process.env.API_URL || 'http://127.0.0.1:8000';

export default {
  api,
  web,
  port,
  axios: args => axios({ api, ...(args || {}) }),
  dev: process.env.NODE_ENV !== 'production',
  sentry: {
    DSN: process.env.SENTRY_DSN,
    publicDSN: process.env.SENTRY_PUBLIC_DSN,
  },
};
