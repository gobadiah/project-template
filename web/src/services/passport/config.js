import config from '~/config';

config.google = {
  clientId: process.env.GOOGLE_CLIENT_ID,
  clientSecret: process.env.GOOGLE_CLIENT_SECRET,
  scope: [
    'profile',
    'email',
    'https://www.googleapis.com/auth/user.birthday.read',
    'https://www.googleapis.com/auth/user.emails.read',
    'https://www.googleapis.com/auth/user.phonenumbers.read',
    'https://www.googleapis.com/auth/userinfo.email',
    'https://www.googleapis.com/auth/userinfo.profile',
  ],
};

config.facebook = {
  clientId: process.env.FACEBOOK_CLIENT_ID,
  clientSecret: process.env.FACEBOOK_CLIENT_SECRET,
};

export default config;
