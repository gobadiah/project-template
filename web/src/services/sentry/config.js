import config from '~/config';

let clientInstance;
let serverInstance;

config.sentry = {
  DSN: process.env.SENTRY_DSN,
  publicDSN: process.env.SENTRY_PUBLIC_DSN,
};

export { config };

export const getRaven = () => {
  const clientSide = typeof window !== 'undefined';

  if (clientSide) {
    if (!clientInstance) {
      const Raven = require('raven-js'); // eslint-disable-line global-require
      Raven.config(config.sentry.publicDSN).install();
      clientInstance = Raven;
    }

    return clientInstance;
  }

  if (!serverInstance) {
    // NOTE: raven (for node) is not bundled by webpack (see rules in next.config.js).
    const RavenNode = require('raven'); // eslint-disable-line global-require
    RavenNode.config(config.sentry.DSN).install();
    serverInstance = RavenNode;
  }
  return serverInstance;
};
