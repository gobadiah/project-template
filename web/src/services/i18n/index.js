import i18n from 'i18next';
import backend from 'i18next-xhr-backend';
import LanguageDetector from 'i18next-browser-languagedetector';

export const availableLanguages = ['fr', 'en'];
export const availableNamespaces = ['common', 'index'];

const detection = {
  order: ['querystring', 'cookie', 'localStorage', 'navigator', 'header'],

  lookupQuerystring: 'lng',
  lookupCookie: 'i18next',
  lookupLocalStorage: 'i18nextLng',

  caches: false,
};

const options = {
  fallbackLng: 'fr',
  load: 'languageOnly',

  ns: availableNamespaces,
  defaultNS: 'common',

  debug: false,

  react: {
    wait: true,
  },

  saveMissing: true,

  detection,

  interpolation: {
    escapeValue: false,
  },
};

if (process.browser) {
  i18n.use(backend)
    .use(LanguageDetector);
  options.caches = ['localStorage', 'cookie'];
  options.backend = {
    loadPath: '/locales/{{lng}}/{{ns}}.json',
    addPath: '/locales/add/{{lng}}/{{ns}}',
  };
}

if (!i18n.isInitialized) {
  i18n.init(options);
}

i18n.getInitialProps = (req, nss) => {
  let namespaces = nss;
  if (typeof namespaces === 'undefined') {
    namespaces = i18n.options.defaultNS;
  }
  if (typeof namespaces === 'string') {
    namespaces = [namespaces];
  }

  req.i18n.toJSON = () => null;

  const initialI18nStore = {};
  req.i18n.languages.forEach((l) => {
    initialI18nStore[l] = {};
    namespaces.forEach((ns) => {
      initialI18nStore[l][ns] = req.i18n.services.resourceStore.data[l] ?
        req.i18n.services.resourceStore.data[l][ns] || {} : {};
    });
  });

  return {
    i18n: req.i18n,
    initialI18nStore,
    initialLanguage: req.i18n.language,
  };
};

export default i18n;
