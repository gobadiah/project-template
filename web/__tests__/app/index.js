/* eslint-disable global-require */

import path from 'path';
import request from 'supertest';

import createApp from '~/app';

describe('App integration', () => {
  let server;

  beforeAll(() => createApp(false).then((value) => { server = value.listen(0); }));

  afterAll(() => { server.close(); });

  it('should respond with 200 status for index and contain some french text', () =>
    request(server)
      .get('/')
      .expect(200)
      .expect(/>Salut le monde !<\/div>/));

  it('should response normally to non-next endpoint (e.g. /api)', () =>
    request(server)
      .get('/api')
      .expect(200)
      .expect('Api endpoint'));
});

describe('App unit', () => {
  it('should use dependencies', () => {
    jest.resetModules();
    const availableLanguages = ['fr'];
    const availableNamespaces = ['common'];
    require('../../src/config/i18n').availableLanguages = availableLanguages;
    require('../../src/config/i18n').availableNamespaces = availableNamespaces;
    const i18n = require('../../src/config/i18n').default;
    i18n.use.mockImplementation(() => i18n);
    i18n.init.mockImplementation(() => {});

    const mockBackend = 'backend';

    jest.mock('../../src/config/i18n');
    jest.mock('i18next-node-fs-backend', () => mockBackend);

    const LanguageDetector = 'LanguageDetector';
    const i18nextMiddleware = require('i18next-express-middleware').default;
    const handle = 'handle';
    const missingKeyHandler = 'missingKeyHandler';
    i18nextMiddleware.handle.mockImplementation(() => handle);
    i18nextMiddleware.missingKeyHandler.mockImplementation(() => missingKeyHandler);
    require('i18next-express-middleware').LanguageDetector = LanguageDetector;

    jest.mock('i18next-express-middleware');

    jest.mock('serve-favicon');
    const favicon = require('serve-favicon');
    favicon.mockImplementation(() => 'favicon');

    const dev = false;

    const app = require('~/app').default;
    const result = app(dev);
    expect(typeof result.then === 'function').toBe(true);

    expect(i18n.use).toHaveBeenCalledTimes(2);
    expect(i18n.use).toHaveBeenCalledWith(mockBackend);
    expect(i18n.use).toHaveBeenCalledWith(LanguageDetector);

    expect(i18n.init).toHaveBeenCalledTimes(1);
    expect(i18n.init.mock.calls[0][0]).toEqual({
      preload: availableLanguages,
      ns: availableNamespaces,
      backend: {
        loadPath: path.join(__dirname, '../../locales/{{lng}}/{{ns}}.json'),
        addPath: path.join(__dirname, '../../locales/{{lng}}/{{ns}}.missing.json'),
        jsonIndent: 2,
      },
    });
    const func = i18n.init.mock.calls[0][1];

    const mockApp = {
      prepare: jest.fn(() => ({
        then: jest.fn(fn => fn()),
      })),
    };

    jest.mock('next', () => jest.fn(() => mockApp));

    const next = require('next');

    const mockHandler = 'handler';
    jest.mock('../../src/routes', () => ({
      getRequestHandler: jest.fn(() => mockHandler),
    }));

    const passport = require('../../src/passport').default;
    jest.mock('../../src/passport');
    passport.mockImplementation(() => {});

    const express = require('express');
    const mockStatic = 'static';
    express.static.mockImplementation(() => mockStatic);
    const mockServer = {
      use: jest.fn(),
      get: jest.fn(),
      post: jest.fn(),
    };
    express.mockImplementation(() => mockServer);

    jest.mock('express');

    func();

    expect(next).toHaveBeenCalledTimes(1);
    expect(next).toHaveBeenCalledWith({ dev, dir: './src' });

    expect(favicon).toHaveBeenCalledTimes(1);
    expect(favicon).toHaveBeenCalledWith(path.join(
      __dirname,
      '../../src/static',
      'favicon.ico',
    ));

    expect(i18nextMiddleware.handle).toHaveBeenCalledTimes(1);
    expect(i18nextMiddleware.handle).toHaveBeenCalledWith(i18n);

    expect(mockServer.use).toHaveBeenCalledTimes(3);
    expect(mockServer.use).toHaveBeenCalledWith('favicon');
    expect(mockServer.use).toHaveBeenCalledWith(handle);
    expect(mockServer.use).toHaveBeenCalledWith('/locales', mockStatic);

    expect(passport).toHaveBeenCalledTimes(1);
    expect(passport).toHaveBeenCalledWith(mockServer);

    expect(mockServer.get).toHaveBeenCalledTimes(2);
    expect(mockServer.get.mock.calls[0][0]).toEqual('/api');
    const func2 = mockServer.get.mock.calls[0][1];
    const res = {};
    res.status = jest.fn(() => res);
    res.send = jest.fn(() => res);
    res.end = jest.fn();

    func2(undefined, res);
    expect(res.status).toHaveBeenCalledTimes(1);
    expect(res.status).toHaveBeenCalledWith(200);
    expect(res.send).toHaveBeenCalledTimes(1);
    expect(res.send).toHaveBeenCalledWith('Api endpoint');
    expect(res.end).toHaveBeenCalledTimes(1);

    expect(mockServer.get).toHaveBeenCalledWith('*', mockHandler);

    expect(i18nextMiddleware.missingKeyHandler).toHaveBeenCalledTimes(1);
    expect(i18nextMiddleware.missingKeyHandler).toHaveBeenCalledWith(i18n);

    expect(mockServer.post).toHaveBeenCalledTimes(1);
    expect(mockServer.post).toHaveBeenCalledWith('/locales/add/:lng/:ns', missingKeyHandler);

    expect(result).resolves.toBe(mockServer);
  });
});
