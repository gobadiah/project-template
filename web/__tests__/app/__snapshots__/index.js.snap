// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App unit should have options matching snapshot 1`] = `
Object {
  "backend": Object {
    "jsonIndent": 2,
  },
  "ns": Array [
    "common",
    "index",
  ],
  "preload": Array [
    "fr",
    "en",
  ],
}
`;
